### PR TITLE
Module 1 Project Exercise

### PR DESCRIPTION
**Pull request by pat-morris**

### CODE REVIEW

File: app.py [hunk 1]
```diff
@@ -1,12 +1,55 @@
 from flask import Flask, render_template, request, redirect, url_for
 import session_items as session
+from flask_modus import Modus
```
Comment (Jonesey13) [inline app.py]: You probably don't want to import this as it doesn't actually change the nature of requests being made to the flask server (e.g. POSTs override to DELETEs will still actually get send as POSTs)

Also this library: 

1. Needs to be added to requirements.txt to work (this needs to run in a virtual environment; see the README file)
1. Relies on an obsolete of werkzeug

File: app.py [hunk 2]
```diff
@@ -1,12 +1,55 @@
...
 def index():
-    return 'Hello World!'
+    return render_template('index.html', get_items=session.get_items())
+
+# Add an item
+@app.route('/add_item', methods=['GET', 'POST'])
+def add_item():
+
+    if request.method == 'POST':
+
```
Comment (Jonesey13) [inline app.py]: Don't really need all these line breaks

File: app.py [hunk 3]
```diff
@@ -1,12 +1,55 @@
...
+        req = request.form
+
+        new_item = req['title']
```
Comment (Jonesey13) [inline app.py]: Probably want to validate against the case where title is empty (i.e. add server side validation along the client side validation you have below)

File: app.py [hunk 4]
```diff
@@ -1,12 +1,55 @@
...
+        return redirect('/')
+
+    return render_template('add_item.html')
+
+@app.route('/read_item/<int:id>', methods=['GET', 'PATCH', 'DELETE'])
+def read_item(id):
+    found_item = session.get_item(id)
+    get_items = session.get_items()
+    if request.method == b'PATCH':
+        #found_item.status = request.form['status']
```
Comment (Jonesey13) [inline app.py]: As we're using source control obsolete code like this can be safely removed

File: app.py [hunk 5]
```diff
@@ -1,12 +1,55 @@
...
+        #found_item.title = request.form['title']
+        return redirect(url_for('index'))
+    if request.method == b'DELETE':
+        session.delete_item(id)
+        return redirect(url_for('index'))
+    return render_template('read_item.html', found_item=found_item)
+
+# Edit an item
```
Comment (Jonesey13) [inline app.py]: Redundant comment

File: session_items.py [hunk 6]
```diff
@@ -67,3 +67,22 @@ def save_item(item):
...
+    """
+    item = get_item(id)
+
+    if item != None:
+        
+        existing_items = get_items()
+
+        existing_items.remove(item)
+
+        save_item(item)
```
Comment (Jonesey13) [inline session_items.py]: This line does nothing as the item has already been deleted

File: templates/delete_item.html [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+{% block title %}Delete To-Do Item{% endblock %}
+
+{% block content %}
+<h1>Delete Item</h1>
+
+{% if get_items %}
+<form action="" method="post" name="item_list">
+    <table class="table is-striped is-fullwidth">
+        {% for item in get_items %}
+        <tr>
```
Comment (Jonesey13) [inline templates/delete_item.html:11]: This should be on a new level of indentation (to show its inside the `{% ... %}` block.

File: app.py [hunk 8]
```diff
@@ -1,12 +1,55 @@
...
+
+    if request.method == 'POST':
+
+        req = request.form
+
+        new_item = req['title']
+
+        print(new_item)
+
+        session.add_item(new_item)
```
Comment (Jonesey13) [inline app.py]: This doesn't update the status

File: templates/edit_item.html [hunk 9]
```diff
@@ -0,0 +1,20 @@
...
+{% block content %}
+<div class="jumbotron">
+  <h1 class="display-4">To-Do App</h1>
+  <p class="lead">Just another to-do app.</p>
+</div>
+
+<div class="row">
+  <div class="col-md-12">
+    <h2>Edit this item:</h2>
+    <form method="POST" action="{{url_for('read_item', id=found_item.id)}}?_method=PATCH">
```
Comment (Jonesey13) [inline templates/edit_item.html:13]: This endpoint doesn't actually update the item

File: templates/edit_item.html [hunk 10]
```diff
@@ -0,0 +1,20 @@
...
+      <label>Status:</label>
+      <input type="text" size="50" name="status" value="{{ found_item.status }}">
+      <label>Title:</label>
+      <input type="text" size="50" name="title" value="{{ found_item.title }}">
```
Comment (Jonesey13) [inline templates/edit_item.html:20]: Needs a `required` attribute

**Review by Jonesey13 — CHANGES_REQUESTED** [review]
Nearly there! Currently the Flask-Modus dependency is preventing this PR from passing this Module (as the code doesn't work in the virtual environment).

Apart from that the core exercise goals work without issue.

I've also left some comments regarding the stretch goals (in particular the statuses don't appear to work).

File: app.py (hunk 1 above)
Comment (pat-morris) [reply app.py]: Thanks, Alex. As it's obsolete, I have uninstalled the flask-modus package and removed references to it in my app.py file. I'll have a look for an alternative.

File: app.py (hunk 2 above)
Comment (pat-morris) [reply app.py]: Agreed! Line breaks have now been removed! Perhaps I should also do this when writing emails ;)

File: session_items.py (hunk 6 above)
Comment (pat-morris) [reply session_items.py]: Doh! That is a good point, at the time it made perfect sense! :)

File: app.py (hunk 8 above)
Comment (pat-morris) [reply app.py]: I've reduced the scope here, initially I was trying to allow users the option of updatating two fields i.e. status and title... this proved difficult! I then realised the stretch goal was only asking to add functionality to mark an item as completed i.e. only one field needed to change! This made the process a lot easier to solve!

File: templates/edit_item.html (hunk 9 above)
Comment (pat-morris) [reply templates/edit_item.html:13]: Ah yes, sorry about that! I uninstall the flask-modus module and made some changes in the app.py and read_item.html. The edit_item.html file is now redundant!

**Pull request merged**